proposal: Reliability of kmesh restart (#417)
kmesh-net/kmesh: docs/proposal/kmesh-reliability.md

---
title: Kmesh reliability
authors:
- "@kwb0523" # Authors' GitHub accounts here.
- "@nlgwcy"
reviewers:
- "@bitcoffeeiux"
- "@hzxuzhonghu"
- "@supercharge-xsy"
approvers:
- "@robot"

creation-date: 2024-06-04

---

## Kmesh reliability

### Summary

This proposal describes how to ensure that the governance capability is not affected when the Kmesh is restarted.

### Motivation

The problem is from [issue_322](https://github.com/kmesh-net/kmesh/issues/322). After the Kmesh is restarted, the pods managed by the Kmesh are no longer taken over by the Kmesh. As a result, the governance function is affected.

#### Goals

In normal and abnormal restart scenarios, the data plane governance function is not affected, including:

> 1. After the pod is restarted, it is automatically restored to the status before the restart.

Reviewer: ```suggestion
1. After kmesh daemon is restarted, it is automatically restored to the status before.
```

> 2. During the Kmesh restart, newly established links have the governance capability.

Reviewer: ```suggestion
2. During the Kmesh restart, newly established connections should be governed by Kmesh.
```

> 3. After the Kmesh is restarted, the newly created link has the governance capability.

Reviewer: ditto

4. The governance capability of old links is not affected by restart.

> 5. When Kmesh is deleted, resources can be completely cleared to avoid residual resources.

Reviewer: ```suggestion
5. After Kmesh is uninstalled, resources should be completely cleared to avoid leaking resources.
```

#### Non-Goals

Currently, this document focuses on the Kmesh daemon restart scenario. The following scenarios are not included in the design scope:

1. Upgrade the Kmesh daemon.

### Proposal

The persistence capability of the bpf map and the configuration restoration capability after the Kmesh restart need to be supported, including:

> 1. When the Kmesh is restarted, the bpf map and bpf prog need to be maintained without being deleted, which is the basis for maintaining the data plane governance capability.

Reviewer: s/maintained/persisted

> 2. After the Kmesh is restarted, the configuration restoration module is added to restore the governance information of the pods managed by the Kmesh on the node.

Reviewer: Assume this is during bootstrap, kmesh populate the bpf map again with all the workload/service info

3. In the scenario where the Kmesh is deleted, the bpf prog and bpf map resources must be cleared in a timely manner to ensure that no resource leakage occurs.

> When the Kmesh is started, the bpf prog and bpf map may already exist on the node. In this case, compatibility processing is required.

Reviewer: can you elaborate on `compatibility processing`

### Design Details

#### Status Quo and Strategy Analysis

> The following uses the workload mode as an example to describe the operations on each BPF map during Kmesh governance:

Reviewer 2: Is it a bpf map or a BPF map. In this article it needs to be standardised

> ![](pics/kmesh-workload-bpf-map.svg)

Reviewer: Cool image

As shown in the preceding figure, these BPF maps are classified by feature as follows:

- Workload Model
  - maps: kmesh_frontend | kmesh_service | kmesh_endpoint | kmesh_backend
  - Data source/Time: When the cluster is changed, the Istiod management and control plane delivers the data.
- Kmesh management
  - maps: map_of_manager

>   - Data source/Time: kmesh-plg, which is triggered when a pod is created or deleted;

Reviewer: kmesh-cni and also we have a controller can do this now

- Waypoint interconnection
  - maps: map_of_dst_info
  - Data source/Time: If ndat needs to be delivered to the waypoint in the active connect phase, the original link information is added to the map_of_dst_info, and the map_of_kmesh_socket sockhash table is added to replace the send/recv hook of the socket (to facilitate the BPF hook in the triggering of the receive and transmit processes). Reads the map_of_dst_info record from the BPF prog in the send phase, constructs meta data, and adds the data to the TCP packet.

Impacts of the Kmesh restart on the preceding tables are as follows:

- During the Kmesh restart:

  | Scenario     | workload model | Kmesh management | Kmesh byPass | Auth | waypoint Interconnection                                     |
  | ------------ | -------------- | ---------------- | ------------ | ---- | ------------------------------------------------------------ |

>   | Existed link | Y              | Y                | Y            | Y    | P(If the connection is complete but the send request has not been sent) |

Reviewer: What does Y and N mean

  | New link     | N              | N                | N            | N    | N                                                            |

  Countermeasures:

  - The bpf prog and map need to be maintained during restart.

> - After the Kmesh is restarted:

Reviewer 2: ```suggestion
- After Kmesh restart:
```

  | Scenario     | workload model | Kmesh management | Kmesh byPass | Auth | waypoint Interconnection |
  | ------------ | -------------- | ---------------- | ------------ | ---- | ------------------------ |
  | Existed link | Y              | Y                | Y            | Y    | Y                        |
  | New link     | Y              | N                | Y            | N    | N                        |

  Countermeasures:

  - The bpf prog and map need to be maintained during restart.
  - After the Kmesh is started, the data in the map needs to be restored and the data needs to be synchronized with the control plane.

>   Note: Currently, Kmesh supports only link-level authentication.

Reviewer 2: ```suggestion
  Note: Currently, Kmesh only supports link-level authentication.
```

#### Solutions

> ![](pics/reliability-arch.svg)

Reviewer: This graph is not shown correctly

- eBPF management

  - To keep the governance capability during the Kmesh restart, you need to place the bpf prog and bpf map pin to the /sys/fs/bpf directory on the node.

>   - During the startup of the Kmesh, the BPF PROG is replaced and the BPF MAP is inherited.

Reviewer: > BPF PROG is replaced

How will this influence the governing prog

- Configure recovery

  - Workload resources

>     - During the startup of the Kmesh, the local bpf map data is restored to the user-mode cache before the Kmesh subscribes to workload model data from the Istiod.

Reviewer: Based on offline discussion, how can user space cache be restored correctly is still not clear

    - Incremental subscription to Istiod based on workload resources in the local cache;
  - Kmesh management/byPass

>     - Move the management logic from kmesh-plg to kmesh-daemon.

Reviewer: kmesh-plg/kmesh-cni ?

    - After the restart, synchronize the configuration based on the information restored by the bpf map and the annotation information on the subscribed pod. (The data needs to be compared by the Kmesh.)
  - Authentication
    - Restore the user-mode authentication cache information based on the authentication rule delivered by the Istiod.
  - Waypoint interconnection

>     - Restore the user-mode authentication cache information based on the authentication rule delivered by the Istiod. The procedure is the same as that for restoring workload resources.

Reviewer: ?

- Resource recycling

  In the scenario where the Kmesh is uninstalled, the bpf prog and bpf map need to be cleared.

  - Postponed clearance

>     When the Kmesh exits, start a delay clearing task on the node. If the Kmesh restarts abnormally, delete the scheduled task after the restart.

Reviewer: ALternatively we can detect whether there is a new kmesh daemon pod in the same node. And only clear the resources if no such pod exists
